### PR TITLE
adds mech drill

### PR DESCRIPTION
**Pull request by Naive817**
## About the PR

Adds mech drill that can be installed on ripley to mine rocks
It cant be used on other things or as a melee weapon as for now.
Also added it into Ripley research entry.

**Media**

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

It doesn't interact with things while in hand, and only interacts with entities with gatherable component, but doesn't function as a melee weapon now because right click would hit the pilot.

https://user-images.githubusercontent.com/31364560/224895403-6792b86e-7a29-45d7-bd9a-6512cdded93f.mp4

**Changelog**

:cl:
- add: Added mech drill

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
needs a screenshot/video of it being tested

**Comment by Naive817** [comment]
> needs a screenshot/video of it being tested

video is added, went lazy when I made the PR

File: Resources/Prototypes/Entities/Objects/Specific/Mech/mecha_equipment.yml [hunk 1]
```diff
@@ -23,4 +23,19 @@
...
   - type: ContainerContainer
     containers:
-      item-container: !type:Container
\ No newline at end of file
+      item-container: !type:Container
+
+- type: entity
+  id: MechEquipmentDrill
+  parent: BaseMechEquipment
+  name: Mecha drill
```
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Objects/Specific/Mech/mecha_equipment.yml:31]:
```suggestion
  name: mecha drill
```

File: Content.Server/Mech/Equipment/Components/MechDrillComponent.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    /// <summary>
+    /// The sound played when a mech is drillbing something
+    /// </summary>
+    [DataField("drillSound")]
+    public SoundSpecifier drillSound = new SoundPathSpecifier("/Audio/Mecha/sound_mecha_hydraulic.ogg");
+
+    public IPlayingAudioStream? AudioStream;
+
+}
```
Comment (mirrorcult) [inline Content.Server/Mech/Equipment/Components/MechDrillComponent.cs:34]:
```suggestion
[RegisterComponent]
public sealed class MechDrillComponent : Component
{
    /// <summary>
    /// The change in energy after each drill.
    /// </summary>
    [DataField("drillEnergyDelta")]
    public float DrillEnergyDelta = -30;

    /// <summary>
    /// How long does it take to drill something?
    /// </summary>
    [DataField("drillDelay")]
    public float DrillDelay = 0.25f;

    /// <summary>
    /// The sound played when a mech is drilling something
    /// </summary>
    [DataField("drillSound")]
    public SoundSpecifier DrillSound = new SoundPathSpecifier("/Audio/Mecha/sound_mecha_hydraulic.ogg");

    public IPlayingAudioStream? AudioStream;
}
```

conventions

File: Content.Server/Mech/Equipment/EntitySystems/MechDrillSystem.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        args.Handled = true;
+        component.AudioStream = _audio.PlayPvs(component.drillSound, uid);
+        _doAfter.DoAfter(new DoAfterEventArgs(args.User, component.drillDelay, target:target, used:uid)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true
+        });
+    }
+
+    private void OnMechdrill(EntityUid uid, MechDrillComponent component, DoAfterEvent args)
```
Comment (mirrorcult) [inline Content.Server/Mech/Equipment/EntitySystems/MechDrillSystem.cs:69]:
```suggestion
    private void OnMechDrill(EntityUid uid, MechDrillComponent component, DoAfterEvent args)
```

File: Content.Server/Mech/Equipment/EntitySystems/MechDrillSystem.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        if (!_interaction.InRangeUnobstructed(args.User, target))
+            return;
+
+        args.Handled = true;
+        component.AudioStream = _audio.PlayPvs(component.drillSound, uid);
+        _doAfter.DoAfter(new DoAfterEventArgs(args.User, component.drillDelay, target:target, used:uid)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true
```
Comment (mirrorcult) [inline Content.Server/Mech/Equipment/EntitySystems/MechDrillSystem.cs:65]: set `RaiseOnTarget` and `RaiseOnUser` = false in the doafter args

File: Content.Server/Mech/Equipment/EntitySystems/MechDrillSystem.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            return;
+
+        if (mech.Energy + component.drillEnergyDelta < 0)
+            return;
+
+        if (!_interaction.InRangeUnobstructed(args.User, target))
+            return;
+
+        args.Handled = true;
+        component.AudioStream = _audio.PlayPvs(component.drillSound, uid);
```
Comment (mirrorcult) [inline Content.Server/Mech/Equipment/EntitySystems/MechDrillSystem.cs:61]: sound could stand to be a little quieter imo, judging from the video

**Review by mirrorcult — CHANGES_REQUESTED** [review]
the interaction with `GatheringTool` here seems really weird? technically two doafters would try to happen at once since mech drill tries to start its own, when really all you need with this interaction is something like:

- before gathering doafters, raise an event to see if it should go through--mechdrill handles it and cancels if not enough energy or etc mech related checks. if it should go through it starts playing the audio
- after doafter (which gets raised by gatherable), do all the usual logic you already have there

**Comment by Naive817** [comment]
> the interaction with `GatheringTool` here seems really weird? technically two doafters would try to happen at once since mech drill tries to start its own, when really all you need with this interaction is something like:
> 
> * before gathering doafters, raise an event to see if it should go through--mechdrill handles it and cancels if not enough energy or etc mech related checks. if it should go through it starts playing the audio
> * after doafter (which gets raised by gatherable), do all the usual logic you already have there

`GatherableSystem` subscribed to `InteractUsingEvent`, which is raised when a target entity is interacted with by a user while holding an object in their hand.
mech systems used `InteractNoHandEvent`, which is cursed according to remarks

**Comment by Naive817** [comment]
> the interaction with `GatheringTool` here seems really weird? technically two doafters would try to happen at once since mech drill tries to start its own, when really all you need with this interaction is something like:
> 
> * before gathering doafters, raise an event to see if it should go through--mechdrill handles it and cancels if not enough energy or etc mech related checks. if it should go through it starts playing the audio
> * after doafter (which gets raised by gatherable), do all the usual logic you already have there

`GatherableSystem` subscribed to `InteractUsingEvent`, which is raised when a target entity is interacted with by a user while holding an object in their hand.
mech systems used `InteractNoHandEvent`, which is cursed according to remarks

**Comment by Naive817** [comment]
> the interaction with `GatheringTool` here seems really weird? technically two doafters would try to happen at once since mech drill tries to start its own, when really all you need with this interaction is something like:
> 
> * before gathering doafters, raise an event to see if it should go through--mechdrill handles it and cancels if not enough energy or etc mech related checks. if it should go through it starts playing the audio
> * after doafter (which gets raised by gatherable), do all the usual logic you already have there

`GatherableSystem` subscribed to `InteractUsingEvent`, which is raised when a target entity is interacted with by a user while holding an object in their hand.
mech systems used `InteractNoHandEvent`, which is cursed according to remarks

File: Content.Server/Mech/Equipment/Components/MechDrillComponent.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Threading;
+using Robust.Shared.Audio;
+
+namespace Content.Server.Mech.Equipment.Components;
+
+/// <summary>
+/// A piece of mech equipment that drills entities and stores them
+/// inside of a container so large objects can be moved.
+/// </summary>
```
Comment (metalgearsloth) [inline Content.Server/Mech/Equipment/Components/MechDrillComponent.cs:9]: Looks like you copied this from mechgrabber.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I'm skeptical of making mech copies of code just for basic interactions, especially when it's not being predicted.

Is there any strong reason why this isn't just using gatherable?

File: Content.Server/Mech/Equipment/Components/MechDrillComponent.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    /// How long does it take to drill something?
+    /// </summary>
+    [DataField("drillDelay")]
+    public float DrillDelay = 0.25f;
+    /// <summary>
+    /// The sound played when a mech is drilling something
+    /// </summary>
+    [DataField("drillSound")]
+    public SoundSpecifier DrillSound = new SoundPathSpecifier("/Audio/Mecha/sound_mecha_hydraulic.ogg");
+    public IPlayingAudioStream? AudioStream;
```
Comment (metalgearsloth) [inline Content.Server/Mech/Equipment/Components/MechDrillComponent.cs:28]: Delay needs to be 1 minimum, 0.25 means the do_after doesn't even show.

**Comment by metalgearsloth** [comment]
@Naive817 Are you coming back to this.

**Comment by PilgrimViis** [comment]
> @Naive817 Are you coming back to this.

I did the same feature for an unofficial server. If you want guys, I could pick up this task and create a new PR in this repo. You can check out a video of how I've implemented it here. https://github.com/Workbench-Team/space-station-14/pull/161/files

**Comment by AJCM-git** [comment]
> > @Naive817 Are you coming back to this.
> 
> I did the same feature for an unofficial server. If you want guys, I could pick up this task and create a new PR in this repo. You can check out a video of how I've implemented it here. [Workbench-Team/space-station-14#161 (files)](https://github.com/Workbench-Team/space-station-14/pull/161/files)

Push your changes to this PR and we will review that if you want to contribute this feature, opening a new one for changes is not advised (worst case you can force pusH)

**Comment by AJCM-git** [comment]
> > @Naive817 Are you coming back to this.
> 
> I did the same feature for an unofficial server. If you want guys, I could pick up this task and create a new PR in this repo. You can check out a video of how I've implemented it here. [Workbench-Team/space-station-14#161 (files)](https://github.com/Workbench-Team/space-station-14/pull/161/files)

Yeah, you can take it (unless @Naive817 says otherwise

**Pull request closed without merging**